src/id_allocator.rs

```diff
@@ -0,0 +1,204 @@
+// Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+
+//! Provides allocation and releasing strategy for IDs.
+//!
+//! This module implements an allocation strategies for all resource types
+//! that can be abstracted to an integer.
+
+use std::collections::BTreeSet;
+use std::fmt;
+use std::result;
+
+/// Error type for IdAllocator usage.
+#[derive(Debug, PartialEq)]
+pub enum Error {
+    /// All ids from the range specified are exhausted.
+    Overflow,
+    /// An id that is not part of the specified range was requested to be released.
+    InvalidInput(u32),
+    /// There are no more IDs available in the manage range
+    ResourceExhausted,
+}
+
+pub type Result = result::Result<u32, Error>;
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use self::Error::*;
+
+        match *self {
+            Overflow => write!(f, "Id counter overflowed."),
+            InvalidInput(id) => write!(f, "Specified id: {} is not in the range.", id),
+            ResourceExhausted => {
+                write!(f, "There are no more available ids in the specified range.")
+            }
+        }
+    }
+}
+
+/// # Example
+///
+/// ```rust
+/// # use vm_allocator::IdAllocator;
+///
+/// let mut legacy_irq_allocator = IdAllocator::new(5, 23);
+///
+/// let irq_nr = legacy_irq_allocator.allocate_id();
+/// if irq_nr.is_err() {
+///     println!("Interrupt number allocation failed: {:?}", irq_nr);
+/// }
+/// let irq = 8;
+/// let freed_id_result = legacy_irq_allocator.free_id(irq);
```

Comment:
What happens if an unallocated id is freed?

Reply:
If an ID that was not allocated before we will return an error. In this [line](https://github.com/rust-vmm/vm-allocator/blob/bde96f123e5d56e285a393d5459e4c0a0f48b634/src/id_allocator.rs#L122) we check to see if the id requested is in the range we are working with. We explicitly check that the ID was allocated `self.next_id < id` if no we return `InvalidInput` for now but I think the error name will change.